### PR TITLE
 Block debugger attach after any ApplyUpdate changes

### PR DESCRIPTION
**Pull request by mikem8361**
DebugActiveProcess now fails if any metadata updates have been applied with the ApplyUpdate API.

Add better error message for ApplyUpdate while debugging.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/Resources/Strings.resx [hunk 1]
```diff
@@ -3685,6 +3685,9 @@
   <data name="InvalidOperation_EditFailed" xml:space="preserve">
     <value>The assembly update failed.</value>
   </data>
+  <data name="NotSupported_DebuggerAttached" xml:space="preserve">
+    <value>Assembly updates not supported while debuggger is attached.</value>
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:3689]: Maybe something like:
```suggestion
    <value>Updates were applied to an assembly in the process. The debugger is unable to attach.</value>
```
?

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:3689]: Oh, wait, nevermind, this error message is about the debugger already being attached?

Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:3689]:
```suggestion
    <value>Assembly updates cannot be applied while a debugger is attached.</value>
```

Comment (mikem8361) [reply src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:3689]: Yes, this change was about the other block.

**Review by jkotas — APPROVED** [review]

**Comment by jkotas** [comment]
cc @lambdageek

**Review by stephentoub — APPROVED** [review]
What is the error message someone will get if they try to attach after an edit?

**Comment by mikem8361** [comment]
The debugger API returns CORDBG_E_DEBUGGING_NOT_POSSIBLE HRESULT when there have been updates.  It is the closest error I could come up with.  VS displays the following message as a result:

"Unable to attach to the process. Debugging is not possible due to an incompatibility within the CLR implementation"

**Comment by jkotas** [comment]
Should we define a new HRESULT for this so that VS can give better error message in this situation?

**Comment by stephentoub** [comment]
Thanks.  What would it take to do better (after this PR) such that the error message is more descriptive about the particular case? Introduce a new error code and teach VS what to say in response?

**Comment by stephentoub** [comment]
(Jan beat me to my question by 45 seconds apparently :-)

**Comment by mikem8361** [comment]
The problem with a new HRESULT is that VS will display something very generic until they do another release.

**Comment by mikem8361** [comment]
Like "Unable to attach. Unknown error: 0x8xxxxxxxx" I'm assuming.

**Comment by stephentoub** [comment]
I think that's ok.  Ideally we could get the error=>string mapping into a near-term 16.xx release, but even if we couldn't, I'd rather have a better experience here for the longer term.  After all, one of the primary purposes of having this block in place is to provide a better user experience that's explanatory of the situation than what would happen if you could successfully attach but find yourself unable to meaningfully debug and not know why.  And if it is "Unable to attach. Unknown error: 0x8xxxxxxxx", worst case that 0x8xxxxxxxx becomes searchable and we can ensure the right information is exposed in docs about what that message means.

**Comment by mikem8361** [comment]
How about:

`#define CORDBG_E_ASSEMBLY_UPDATES_APPLIED EMAKEHR(0x1c69)`

The resulting VS message is:

"Unable to attach to the process. Operation not supported. Unknown error: 0x80131c69"

File: src/coreclr/pal/prebuilt/inc/corerror.h [hunk 2]
```diff
@@ -383,6 +383,7 @@
 #define CORDBG_E_DATA_TARGET_ERROR EMAKEHR(0x1c61)
 #define CORDBG_E_NO_IMAGE_AVAILABLE EMAKEHR(0x1c64)
 #define CORDBG_E_UNSUPPORTED_DELEGATE EMAKEHR(0x1c68)
+#define CORDBG_E_ASSEMBLY_UPDATES_APPLIED EMAKEHR(0x1c69)
```
Comment (jkotas) [inline src/coreclr/pal/prebuilt/inc/corerror.h:386]: Could you please also add it to src\coreclr\inc\corerror.xml ?

**Comment by mikem8361** [comment]
I think I need help with the message. I was trying to keep the error and text not specific to attach just in case we need to use it in some other case.

```
<HRESULT NumericValue="0x80131c69">
  <SymbolicName>CORDBG_E_ASSEMBLY_UPDATES_APPLIED</SymbolicName>
  <Message>"The operation is not supported because assembly updates have been applied."</Message>
  <Comment>The operation is not supported because assembly updates have been applied.</Comment>
</HRESULT>
```

**Comment by stephentoub** [comment]
> "Unable to attach to the process. Operation not supported. Unknown error: 0x80131c69"

LGTM, thanks.  Then we'll just need to get a dedicated, descriptive message added into VS for this new error code for when someone's able to use that newer version.

**Pull request merged**